### PR TITLE
Update config_list_from_json function to support YAML parsing

### PR DESCRIPTION
**Pull request by Mai0313**
## Why are these changes needed?

Add Yaml file supported for OAI_CONFIG_LIST by simply changing from `json.load` to `yaml.safe_load`

## Related issue number

#2559

## Checks

- [X] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by Mai0313** [comment]
> @microsoft-github-policy-service agree

@microsoft-github-policy-service agree company="Mediatek"

**Comment by Mai0313** [comment]
Idk why there is no yaml module, i thought it was a default module for Python

**Review by ekzhu — CHANGES_REQUESTED** [review]
Thanks for the PR! Could you also update the documentation: website/docs/topics/llm_configuration.ipynb

**Comment by Mai0313** [comment]
> Thanks for the PR! Could you also update the documentation: website/docs/topics/llm_configuration.ipynb

ofc, i will update it once i am at work!

**Comment by Mai0313** [comment]
> website/docs/topics/llm_configuration.ipyn

oh sorry, it is my first time to use `mdx` format, is there anyway I can preview the changes before commit?
I tried [this extension](https://marketplace.visualstudio.com/items?itemName=xyc.vscode-mdx-preview), but it seems not working for me.

**Comment by Mai0313** [comment]
@ekzhu 
It is my first time to use `mdx` format and I do not know where I can preview the changes for `mdx`, so basically, the way I have done is copy-paste the previous content and changed to yaml version.

BTW, I have added `pyyaml` to `setup.py` to make sure no one get error when loading config, since we don't use `json` module but `yaml`.


Additionally, I have implemented pytest for the YAML version. Specifically, I utilized `pytest.mark.parametrize` to minimize code duplication.

**Comment by ekzhu** [comment]
> It is my first time to use mdx format and I do not know where I can preview the changes for mdx, so basically, the way I have done is copy-paste the previous content and changed to yaml version.

See https://microsoft.github.io/autogen/docs/contributor-guide/documentation, the `process_notebooks.py` script will generate the MDX files.

File: setup.py [hunk 1]
```diff
@@ -25,6 +25,7 @@
     # Disallowing 2.6.0 can be removed when this is fixed https://github.com/pydantic/pydantic/issues/8705
     "pydantic>=1.10,<3,!=2.6.0",  # could be both V1 and V2
     "docker",
+    "pyyaml",
```
Comment (ekzhu) [inline setup.py:28]: We can add `pyyaml` to test dependencies. But for install dependency let's move this to an optional dependency like those below.

File: autogen/oai/openai_utils.py [hunk 2]
```diff
@@ -504,11 +522,11 @@ def config_list_from_json(
         if os.path.exists(env_str):
             # It is a file location, and we need to load the json from the file.
             with open(env_str, "r") as file:
-                json_str = file.read()
+                json_or_yaml_str = file.read()
         else:
             # Else, it should be a JSON string by itself.
-            json_str = env_str
-        config_list = json.loads(json_str)
+            json_or_yaml_str = env_str
```
Comment (ekzhu) [inline autogen/oai/openai_utils.py:528]: If the yaml dependency is not installed, we can fall back to use `json` module. We can do this by checking for yaml import error during import, and use that information to choose whether to use yaml load or json load.

File: autogen/oai/openai_utils.py [hunk 3]
```diff
@@ -475,21 +476,38 @@ def config_list_from_json(
 
     Args:
         env_or_file (str): The name of the environment variable, the filename, or the environment variable of the filename
-            that containing the JSON data.
+            that containing the JSON/YAML data.
```
Comment (ekzhu) [inline autogen/oai/openai_utils.py:479]: Add explanation for installing `pyyaml`.

File: autogen/oai/openai_utils.py (hunk 2 above)
Comment (Mai0313) [reply autogen/oai/openai_utils.py:528]: oh, it sounds good

File: setup.py (hunk 1 above)
Comment (Mai0313) [reply setup.py:32]: actually I dont see any of test dependency or optional dependency under setup.py
should I put it under test from extra_require?

File: autogen/oai/openai_utils.py [hunk 4]
```diff
@@ -503,11 +530,11 @@ def config_list_from_json(
         if os.path.exists(env_str):
             # It is a file location, and we need to load the json from the file.
```
Comment (thinkall) [inline autogen/oai/openai_utils.py:531]:
```suggestion
            # It is a file location, and we need to load the json/yaml from the file.
```

File: autogen/oai/openai_utils.py [hunk 5]
```diff
@@ -503,11 +530,11 @@ def config_list_from_json(
...
             with open(env_str, "r") as file:
-                json_str = file.read()
+                json_or_yaml_str = file.read()
         else:
             # Else, it should be a JSON string by itself.
```
Comment (thinkall) [inline autogen/oai/openai_utils.py:535]:
```suggestion
            # Else, it should be a JSON/YAML string by itself.
```

File: setup.py (hunk 1 above)
Comment (thinkall) [reply setup.py:32]: You can add it like here: https://github.com/microsoft/autogen/blob/82903f5f89a4f514b267911e56c3032b339f2581/setup.py#L60-L70

**Review by thinkall — COMMENTED** [review]
Thank you very much for the PR, @Mai0313 . I've left a few comments, but it overall looks good to me.

**Comment by thinkall** [comment]
Some tests failed,  could you fix them? For documentation, you can follow https://microsoft.github.io/autogen/docs/contributor-guide/documentation to check it locally.

**Pull request closed without merging**